blackbox/bwc/src/tests.py

```diff
@@ -0,0 +1,162 @@
+"""
+Module that contains basic bwc tests
+"""
+
+import unittest
+import tempfile
+import shutil
+import time
+import os
+import glob
+from typing import NamedTuple
+from io import BytesIO
+from testutils.paths import project_root
+from crate.client import connect
+
+# Uses CrateNode from cr8 instead of crate.testing.layer CrateLayer
+# Because CrateLayer doesn't support customizing path.data behaviour and
+# also doesn't support launching nodes version < 1.0.0
+from cr8.run_crate import CrateNode, get_crate
+
+
+class VersionDef(NamedTuple):
+    version: str
+    upgrade_segments: bool
+
+
+CURRENT_TARBALL = next(iter(
+    glob.glob(os.path.join(project_root, 'app', 'build', 'distributions', '*.tar.gz'))))
+VERSIONS = (
+    VersionDef('1.0.x', False),
+    VersionDef('1.1.x', True),
+    VersionDef('2.0.x', False),
+    VersionDef(CURRENT_TARBALL, False)
+)
+TABLE_WITH_MOST_TYPES = '''
+create table t1 (
+    id int primary key,
+    name string,
+    f float,
+    d double,
+    point geo_point,
+    shape geo_shape,
+    ip_address ip,
+    ts timestamp
+) clustered into 2 shards with (number_of_replicas = 0)
+'''
+
+INSERT = '''
+INSERT INTO t1
+    (id, name, f, d, point, shape, ip_address, ts)
+VALUES
+    (?, ?, ?, ?, ?, ?, ?, ?)
+'''
+
+# Use statements that use different code paths to retrieve the values
+SELECT_STATEMENTS = (
+    'select * from t1',
+    'select distinct id, name, f, d, ip_address, ts from t1',
+    'select * from t1 where id = 1'
+)
+
+
+DUMMY_ROW = (
+    1,
+    'Arthur',
+    238.4,
+    838.42,
+    [29.3, 21.1],
+    'polygon (( 10 10, 10 20, 20 20, 20 15, 10 10))',
+    '127.0.0.1',
+    int(time.time() * 1000)
+)
+
+
+def wait_for_active_shards(cursor):
+    waited = 0
+    duration = 0.01
+    while waited < 20:
+        cursor.execute("select count(*) from sys.shards where state = 'STARTED'")
+        if int(cursor.fetchone()[0]) == 4:
+            return
+        time.sleep(duration)
+        waited += duration
+        duration *= 2
+    raise TimeoutError("Shards didn't become active in time")
+
+
+def run_selects(c, blob_container, digest):
+    for stmt in SELECT_STATEMENTS:
+        print('  ', stmt)
+        c.execute(stmt)
+    print('   Getting blob: ', digest)
+    blob_container.get(digest)
+
+
+class BwcTest(unittest.TestCase):
+
+    def setUp(self):
+        self._path_data = tempfile.mkdtemp()
+        print(f'data path: {self._path_data}')
+        self._on_stop = []
+
+        def new_node(version):
```

Comment:
self._path_data and self._on_stop  is accessed within the function. I could also move the function to the top level and bind the parameters using partial, but thought this is easier.